### PR TITLE
Adding Kraken2 unique kmer counting

### PR DESCRIPTION
**Pull request by maxibor**
Adding the unique Kmer count that has been [ported](https://github.com/DerrickWood/kraken2/pull/249) from [KrakenUniq](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1568-0) into Kraken2 since v2.1.0.

This is quite useful for checking breadth of coverage and read stacking.

It now reports the kmer duplication as well in an extra result table.
_kmer duplication = number of kmer per taxon/ number of unique kmer per taxon_
## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] Make sure your code lints (`nf-core lint .`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [x] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.

### CODE REVIEW

**Comment by maxibor** [comment]
The CI might fail if no new docker image is rebuilt, as the Kraken2 version has been updated from `2.0.9beta` to `2.1.1`

**Comment by ewels** [comment]
The CI should be clever enough to build a new image if the `Dockerfile` or `environment.yml` file has been edited in a PR now:

https://github.com/nf-core/eager/blob/29b6e147bd9a848f84e6d26706af7f652f2165ca/.github/workflows/ci.yml#L28-L37

You can see this happening in the tests: https://github.com/nf-core/eager/pull/687/checks?check_run_id=1955010872#step:4:1

**Review by apeltzer — APPROVED** [review]
Anything we should also add in the JSON Schema maybe? Not sure, no new parameters but the behaviour of the kraken2 run changed - can you maybe check quickly?

**Comment by maxibor** [comment]
No new parameter to add in the JSON Schema: it's an extra in Kraken2 that doesn't affect it otherwise.
Side note, this disting kmer counting should also be reported in the next release of MultiQC https://github.com/ewels/MultiQC/pull/1380

**Comment by apeltzer** [comment]
Looks good to me then - wondering whether this will break MultiQC until then if someone else experiences issues in the current v1.9?

**Comment by ewels** [comment]
Hoping to push a MultiQC release on Friday. Please DM me on Slack to get your PR through if it's needed here 👍🏻

**Comment by maxibor** [comment]
@apeltzer Parts of this PR won't be needed anymore if MultiQC is updated to 1.10 in Eager

**Comment by apeltzer** [comment]
> @apeltzer Parts of this PR won't be needed anymore if MultiQC is updated to 1.10 in Eager

Ok, so maybe we can make sure this ends up in MQC 1.10 and then push for a new eager release that contains an update to MQC 1.10 then too? That should make things align nicely? :-)

**Comment by jfy133** [comment]
I JUST GOT INTERNET! So yes, I think a (roughly) synced release would be a good idea! I have some other fixes to PR (that I worked on offline while on endless trains), which I'll push tomorrow. now I just need a kitchen :sweat_smile:

File: main.nf [hunk 1]
```diff
@@ -3136,15 +3136,33 @@ process kraken {
 
   output:
   file "*.kraken.out" into ch_kraken_out
-  tuple prefix, path("*.kreport") into ch_kraken_report, ch_kraken_for_multiqc
+  tuple prefix, path("*.kraken2_report") into ch_kraken_report, ch_kraken_report_backward_compatibility
```
Comment (jfy133) [inline main.nf:3139]: @maxibor would waiting until MQC 1.10 is release remove the need for this?

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ### `Added`
 
+- [#687](https://github.com/nf-core/eager/pull/687) - Adding Kraken2 unique kmer counting
```
Comment (jfy133) [inline CHANGELOG.md:10]:
```suggestion
- [#687](https://github.com/nf-core/eager/pull/687) - Adds Kraken2 unique kmer counting report
```

File: docs/output.md [hunk 3]
```diff
@@ -669,7 +669,7 @@ Each module has it's own output directory which sit alongside the `MultiQC/` dir
...
 - `metagenomic_classification/` - this contains the output for a given metagenomic classifier.
   - Running MALT will contain RMA6 files that can be loaded into MEGAN6 or MaltExtract for phylogenetic visualisation of read taxonomic assignments and aDNA characteristics respectively. Additional a `malt.log` file is provided which gives additional information such as run-time, memory usage and per-sample statistics of numbers of alignments with taxonomic assignment etc. This will also include gzip SAM files if requested.
-  - Running kraken will contain the Kraken output and report files, as well as a merged Taxon count table.
+  - Running kraken will contain the Kraken output and report files, as well as a merged Taxon count table. You will also get a Kraken kmer duplication table, in a [KrakenUniq](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1568-0) fashion. This is very useful to check for breadth of coverage and detect read stacking. A small number of aligned reads (low coverage) and a kmer duplication >>1 is usually a sign of read stacking, usually synonym of a false positive hit. *Kmer duplication is defined as: number of kmers / number of unique kmers*
```
Comment (jfy133) [inline docs/output.md:672]:
```suggestion
  - Running kraken will contain the Kraken output and report files, as well as a merged Taxon count table. You will also get a Kraken kmer duplication table, in a [KrakenUniq](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1568-0) fashion. This is very useful to check for breadth of coverage and detect read stacking. A small number of aligned reads (low coverage) and a kmer duplication >1 is usually a sign of read stacking, usually indicative of a false positive hit (e.g. from over-amplified libraries). *Kmer duplication is defined as: number of kmers / number of unique kmers*
```

File: main.nf (hunk 1 above)
Comment (apeltzer) [reply main.nf:3139]: Pretty sure it would ==> rather would try pushing this for next release in MQC 1.10 :-)

**Comment by apeltzer** [comment]
I think we could safely rely on this https://github.com/ewels/MultiQC/pull/1380 now - so can assume MQC does read the file, so no need for any custom stuff anymore. 

As we will make a release of the pipeline most likely after MQC 1.10 release, we then only would need to bump versions in the `environment.yaml` once the new release is out and test quickly.

**Comment by apeltzer** [comment]
After #691 is merged, you could make that a bit easier 👍🏻

**Comment by jfy133** [comment]
@maxibor the 1.10 update is now in the `environmental.yml`, so ping me when you've removed the custom/manual stuff and I can re-review (i.e. probably just rubber stamp)

**Comment by jfy133** [comment]
@maxibor you can also get upstream `dev` changes, so you don't get HOPS-related MultiQC related crashes

**Review by apeltzer — APPROVED** [review]
Looks good now & MQC 1.10 supported :-)

**Review by jfy133 — APPROVED** [review]

**Review by apeltzer — APPROVED** [review]

**Pull request merged**